Cart sprint19

FierySasquatch168 opened the pull request:
Архитектура - MVVM
Верстка - В коде
Доска - https://trello.com/b/jyPeNgzT/nftstore
Md-файл - https://github.com/FierySasquatch168/iOS-FakeNFT-StarterProject-Public/tree/cart_sprint19/FakeNFT/Decomposition

CartEpic created using MVVM-design pattern.
1. CartMainScreen created
    - Navigation established using Coordinator pattern and Router, Modules factory responsible for DI
    - Tabbar customization appears in MainTabBarController in Modules folder
    - Tableview with diffable data source applied
    - TableViewCell and CellViewModel created
    - InfoBlock added
    - Navigation to deleteItemScreen through CartCoordinator
    - Navigation to paymentMethodScreen through CartCoordinator
    - Navigation to filterScreen through CartCoordinator
    - EmptyState logic applied based on info from DataStorage 
2. DeleteItemScreen created
    - DeleteItemScreen - custom UIViewController presented overCurrentContext through Cordinator
    - DeleteButton logic uses id from cell via delegate pattern
    - BackButton logic implemented by dismissVC method in Router
3. PaymentMethodScreen created
    - CustomizedNavigationBar
    - CollectionView with diffableDataSource
    - CollectionViewCell with cellViewModel
    - InfoBlock as a stackView with link to WebView with UserLicense 
    - PayButton generating a request to the networkService from ViewModel
4. Create PaymentResultScreen
    - SuccessCase, failCase and loadingCase implemented as cases of an Enum
    - Action button logic (back or retry) is established depending on the case
5. Create SortScreen
   - Screen as alert created with .actionSheet style
   - SortingLogic implemented in DataStore with calls from viewModel

x128 committed:
Added ability to make PUT requests with DefaultNetworkClient

FierySasquatch168 committed:
Create swiftLint.yml

FierySasquatch168 committed:
Delete swiftLint.yml

x128 committed:
Updated Readme and API

FierySasquatch168 committed:
1. Добавлен паттерн Координатор
2. Добавлена фабрика координаторов под каждый модуль
3. Добавлена фабрика модулей для использования инъекции зависимостей и создания экранов
4. Добавлена фабрика навигационных контроллеров для добавления экранов в стек навигации с нужными данными для иконки таббара
5. Добавлен роутер для навигации между экранами, делегатом роутера назначен ScreenDelegate
6. Добавлен основной модуль - ТаббарКонтроллер с собственным координатором и enum для заполнения необходимых ему данных
7. Добавлен модуль категорий с собственным координатором для обработки событий модуля и вьюконтроллер к нему для проверки того, что все работает
8. Добавлен модуль корзина с собственным координатором для обработки событий модуля и вьюконтроллер к нему для проверки того, что все работает
9. Добавлен файл констант для избегания использования строк в коде
10. В дизайн систему (Colors) добавлена часть цветов из UI-Kit фигмы
11. Удален сториборд и ланч-скрин, настройка стартовых вводных происходит в SceneDelegate через запуск основного координатора
12. Внесены изменения в info.plist и таргет проекта (удалены упоминания о сториборде с именем "Main")
13. В ассеты добавлены цвета, которые адаптированы под смену светлой и темной тем

FierySasquatch168 committed:
Добавлена ландшафтная ориентация для внешнего вида таббара

FierySasquatch168 committed:
navigation bar created
table view with diffable data source created
tableView cell created
infoBlock created

FierySasquatch168 committed:
navigation bar created
table view with diffable data source created
tableView cell created
infoBlock created

FierySasquatch168 committed:
masked corners applied to the info block stack view

FierySasquatch168 committed:
masked corners applied to the info block stack view

FierySasquatch168 committed:
added pushViewController method to router
added assets
applied presetable and coordinatable protocols in modules factory to catch events from screens

FierySasquatch168 committed:
added pushViewController method to router
added assets
applied presetable and coordinatable protocols in modules factory to catch events from screens

FierySasquatch168 committed:
cartViewController UI is now made of single stackView

FierySasquatch168 committed:
cartViewController UI is now made of single stackView

FierySasquatch168 committed:
added filter alertController when filterButton tapped on CartViewController
added createFilterAlert() method to CartCoordinator
added makeCartFilterScreenView() method to ModulesFactory
removed navigationTitle from NavigationControllerFactory
applied protocol-oriented style to Router Class (replaced UIViewController with Presentable)

FierySasquatch168 committed:
added filter alertController when filterButton tapped on CartViewController
added createFilterAlert() method to CartCoordinator
added makeCartFilterScreenView() method to ModulesFactory
removed navigationTitle from NavigationControllerFactory
applied protocol-oriented style to Router Class (replaced UIViewController with Presentable)

FierySasquatch168 committed:
alertConstructor implemented for AlertPresenting

FierySasquatch168 committed:
alertConstructor implemented for AlertPresenting

FierySasquatch168 committed:
style property added to CartFilterEnum

FierySasquatch168 committed:
style property added to CartFilterEnum

FierySasquatch168 committed:
onProceed renamed to onFilter

FierySasquatch168 committed:
onProceed renamed to onFilter

FierySasquatch168 committed:
CartTableViewCell now uses MVVM pattern with Combine framework

FierySasquatch168 committed:
CartTableViewCell now uses MVVM pattern with Combine framework

FierySasquatch168 committed:
small code changes

FierySasquatch168 committed:
small code changes

FierySasquatch168 committed:
CartDeleteItemViewController set up.
Separate DataSource storage created, MVVM pattern applied together with Combine framework
Combine framework applied for Cart and CartDeleteItem viewModels

FierySasquatch168 committed:
CartDeleteItemViewController set up.
Separate DataSource storage created, MVVM pattern applied together with Combine framework
Combine framework applied for Cart and CartDeleteItem viewModels

FierySasquatch168 committed:
Combine framework applied for viewModel updates in CartViewController

FierySasquatch168 committed:
Combine framework applied for viewModel updates in CartViewController

FierySasquatch168 committed:
items count and total amount to pay connected to DataStore

FierySasquatch168 committed:
items count and total amount to pay connected to DataStore

FierySasquatch168 committed:
PaymentMethod screen UI done

FierySasquatch168 committed:
PaymentMethod screen UI done

FierySasquatch168 committed:
popToRootViewController method added to Router

FierySasquatch168 committed:
popToRootViewController method added to Router

FierySasquatch168 committed:
settled up navigation to webView
WebViewController set up and works

FierySasquatch168 committed:
settled up navigation to webView
WebViewController set up and works

FierySasquatch168 committed:
tabBar backgroundColor set to match the design

FierySasquatch168 committed:
tabBar backgroundColor set to match the design

FierySasquatch168 committed:
CartPaymentMethodCell selection added

FierySasquatch168 committed:
CartPaymentMethodCell selection added

FierySasquatch168 committed:
filtering of cartItems added

FierySasquatch168 committed:
filtering of cartItems added

FierySasquatch168 committed:
sortDescriptor moved from CartViewModel to DataStore
sorting does not drop after item is deleted - updates are now sent with chosen sort descriptor in DataStore

FierySasquatch168 committed:
sortDescriptor moved from CartViewModel to DataStore
sorting does not drop after item is deleted - updates are now sent with chosen sort descriptor in DataStore

FierySasquatch168 committed:
PaymentResultScreen created
screen UI settled
ViewModel created and connected

FierySasquatch168 committed:
PaymentResultScreen created
screen UI settled
ViewModel created and connected

FierySasquatch168 committed:
CartPaymentResultVC buttons logic updated

FierySasquatch168 committed:
CartPaymentResultVC buttons logic updated

FierySasquatch168 committed:
small code correction

FierySasquatch168 committed:
small code correction

FierySasquatch168 committed:
payment methods loading connected to API

FierySasquatch168 committed:
payment methods loading connected to API

FierySasquatch168 committed:
added RequestConstructor for creation of requests with static methods

FierySasquatch168 committed:
added RequestConstructor for creation of requests with static methods

FierySasquatch168 committed:
added decomposition for cart epic

FierySasquatch168 committed:
added decomposition for cart epic

FierySasquatch168 committed:
Request moved to separate file

FierySasquatch168 committed:
Request moved to separate file

FierySasquatch168 committed:
static func constructWebViewRequest added

FierySasquatch168 committed:
static func constructWebViewRequest added

FierySasquatch168 committed:
helper deleted in favor of using the existing RequestConstructor

FierySasquatch168 committed:
helper deleted in favor of using the existing RequestConstructor

FierySasquatch168 committed:
excess WebViewRequestHandler removed
Modules factory updated (removed initialization of webViewRequestHandler

FierySasquatch168 committed:
excess WebViewRequestHandler removed
Modules factory updated (removed initialization of webViewRequestHandler

FierySasquatch168 committed:
added mockRequest to server for appending items to cart
mockRequest is sent from CartViewModel
removed excessable logic from DataStore
added NftCollections model and NftSingleCollection model for data loading in category module
added networkClient as a initialized property to makeCartScreenView method of ModulesFactory

FierySasquatch168 committed:
added mockRequest to server for appending items to cart
mockRequest is sent from CartViewModel
removed excessable logic from DataStore
added NftCollections model and NftSingleCollection model for data loading in category module
added networkClient as a initialized property to makeCartScreenView method of ModulesFactory

FierySasquatch168 committed:
disabled cellSelection style in CartMainScreen - CartViewController
fixed image showing in CartDeleteItemViewController

FierySasquatch168 committed:
disabled cellSelection style in CartMainScreen - CartViewController
fixed image showing in CartDeleteItemViewController

FierySasquatch168 committed:
moved visibleRows updated to single method in CartViewModel

FierySasquatch168 committed:
moved visibleRows updated to single method in CartViewModel

MaxCode2023 commented:
На iphone 8 немного обрезаются ячейки сверху

![Снимок экрана 2023-06-27 в 09 41 58](https://github.com/FierySasquatch168/iOS-FakeNFT-StarterProject-Public/assets/113161482/2726a571-5e15-4dbb-bd42-e453f1ca7b4f)

MaxCode2023 commented:
Различается картинка на экране с ошибкой, не знаю по макетам она обязательна или нет)

![Снимок экрана 2023-06-27 в 09 44 48](https://github.com/FierySasquatch168/iOS-FakeNFT-StarterProject-Public/assets/113161482/9dd9129c-60d7-4f33-a1a8-bda7e086c806)

MaxCode2023 commented on FakeNFT/Modules/CartModule/Screens/CartDeleteItemScreen/ViewModel/CartDeleteViewModel.swift:
```diff
@@ -0,0 +1,27 @@
+import Foundation
+
+final class CartDeleteViewModel {
+    @Observable private (set) var itemToDelete: CartRow?
```
Заметил, что в других viewModel используется combine, думаю можно здесь заменить реализацию тоже на combine и удалить  класс с @Observable property wrapper

FierySasquatch168 replied:
Да, что-то забыл исправить, спасибо!

MaxCode2023 commented on FakeNFT/Foundation/NetworkClient/RequestConstructor.swift:
```diff
@@ -0,0 +1,36 @@
+
+import Foundation
+
+struct RequestConstructor {
+    static func constructCurrencyRequest(method: HttpMethod, params: [String : String]? = nil) -> NetworkRequest {
+        return Request(
+            endpoint: URL(string: K.Links.apiLink + K.EndPoints.currencies),
+            queryParameters: params,
+            httpMethod: method)
+    }
+    
+    static func constructOrdersRequest(method: HttpMethod) -> NetworkRequest {
+        return Request(
+            endpoint: URL(string: K.Links.apiLink + K.EndPoints.orders),
+            queryParameters: nil,
+            httpMethod: method)
+    }
+    
+    static func constructNftCollectionRequest(method: HttpMethod) -> NetworkRequest {
+        return Request(
+            endpoint: URL(string: K.Links.apiLink + K.EndPoints.singleCollection),
+            queryParameters: ["id" : "1"],
+            httpMethod: method)
+    }
+    
+    static func constructWebViewRequest() -> URLRequest? {
```
Предлагаю у этой структуры сделать приватный init, сейчас ее можно по ошибке инициализировать и после не иметь доступа к методам

FierySasquatch168 replied:
Классная идея, согласен

FierySasquatch168 committed:
added private init to RequestConstructor to prevent unexpected initialisation
applied Combine framework for DeleteItemViewController
@Observable wrapper removed

FierySasquatch168 committed:
added private init to RequestConstructor to prevent unexpected initialisation
applied Combine framework for DeleteItemViewController
@Observable wrapper removed

FierySasquatch168 committed:
fixed problem with rating of the coins
changed RateStarButton to RateStarImageView

FierySasquatch168 committed:
fixed problem with rating of the coins
changed RateStarButton to RateStarImageView

FierySasquatch168 committed:
replaced custom CartRow model by NftSingleCollection model fetched from the server

FierySasquatch168 committed:
replaced custom CartRow model by NftSingleCollection model fetched from the server

FierySasquatch168 committed:
updated cartPlan

FierySasquatch168 committed:
updated cartPlan

FierySasquatch168 committed:
Merge remote-tracking branch 'refs/remotes/origin/cart_sprint19'

FierySasquatch168 committed:
Request updated due to .main branch changes

FierySasquatch168 committed:
updated decomposition.md

FierySasquatch168 committed:
updated decomposition.md

dmisuvorov commented on FakeNFT/Modules/CartModule/Screens/WebViewScreen/WebViewController.swift:
```diff
@@ -0,0 +1,114 @@
+             changeHandler: { [weak self] _, _ in
+                 guard let self else { return }
+                 self.didUpdateProgressValue(webView.estimatedProgress)
+             })
```
А здесь не нужен перед webView self, у меня xcode ругается при сборке, требует
self.didUpdateProgressValue(self.webView.estimatedProgress)

FierySasquatch168 replied:
@dmisuvorov Если есть guard self в замыкании, то компилятор self уже не требует - я просто ленюсь потом подчищать self)

FierySasquatch168 committed:
created custom animated UIView for RequestResult state after payButton tapped

FierySasquatch168 committed:
removed resultImageView property from PaymentResultViewController

FierySasquatch168 committed:
updated constraints of the CartPaymentResultViewController to match design

FierySasquatch168 committed:
removed stackView from CartPaymentResultViewController

FierySasquatch168 committed:
added alert for CartViewModel .failure case of the load() method

FierySasquatch168 committed:
fixed layout issues on CartMainScreen
replaced UILabel for UITextView for result message on CartPaymentResultVC

FierySasquatch168 committed:
fixed layout issues on CartDeleteItemVC using dynamic top and bottom constraints and textViewHeight greaterThan constant

FierySasquatch168 committed:
adjusted vertical spacing between resultTextView and resultView

FierySasquatch168 committed:
added cancellation of subscriptions to viewWillDisappear in all ViewControllers
moved bind() methods to viewWillAppear in all ViewControllers
fixed memory leak in CartCoordinator

FierySasquatch168 committed:
removed didSet from cancellable in CollectionViewCell

FierySasquatch168 committed:
created new method for RequestConstructor
changed paymentRequest to depend on currency
added Decodable struct RequestResult for response decoding
fixed logic of pay() method in CartPaymentResultViewModel
disable ProceedButton in CartPaymentMethodVC before paymentMethod is selected
enable ProceedButton after payment method is selected

FierySasquatch168 committed:
removed unused method from CartPaymentResultScreen

FierySasquatch168 committed:
trailingSwipe delete option added to CartMainScreen

FierySasquatch168 committed:
added loading indicator logic and UI for cartViewController based on CustomAnimatedView usage

FierySasquatch168 committed:
changed size or the animatedView in CartPaymentResultViewController to 75x75

FierySasquatch168 committed:
small code update in CartPaymentResultViewController

FierySasquatch168 committed:
removed delegatePatter from CartModule
added CatalogMainScreen
extended TableViewDataSourceManager

FierySasquatch168 committed:
Kingfisher removed to UIImageViewExtension
CatalogMainScreen connected to tableViewDataSource

FierySasquatch168 committed:
added catalog sorting logic
fixed catalogTableView UI

FierySasquatch168 committed:
catalogCollectionViewController added

Helvetii951 commented on FakeNFT/Constants/K.swift:
```diff
@@ -21,4 +29,17 @@ struct K {
         static let cartTitle = NSLocalizedString("Корзина", comment: "")
         static let statisticsTitle = NSLocalizedString("Статистика", comment: "")
     }
```
Рекомендую использовать библиотеки генерации ресурсов вроде [SwiftGen](https://github.com/SwiftGen/SwiftGen). Понимаю, что нужно было договориться с командой. Но в будущих проектах очень пригодится.

Helvetii951 commented on FakeNFT/Foundation/NetworkClient/RequestConstructor.swift:
```diff
@@ -0,0 +1,44 @@
+
+struct RequestConstructor {
+    
+    private init() {}
```
Приватный инит здесь не совсем нужен. Все-равно все функции статические.

Helvetii951 commented on FakeNFT/Helpers/CustomViews/CustomActionButton.swift:
```diff
@@ -0,0 +1,64 @@
+        setAppearance(for: appearance)
+        setTitle(title, for: .normal)
+        
+        if let cornerRadius {
+            layer.cornerRadius = cornerRadius
+        } else {
+            layer.cornerRadius = 16
+        }
```
Можно переписать, использую [nil coalesing ](https://sarunw.com/posts/what-does-nil-coalescing-operator-means-in-swift/)
```
layer.cornerRadius = cornerRadius ?? 16
```

Helvetii951 commented on FakeNFT/Helpers/CustomViews/CustomActionButton.swift:
```diff
@@ -0,0 +1,64 @@
+            layer.cornerRadius = 16
+        }
+        
+        if let fontWeight {
+            titleLabel?.font = UIFont.systemFont(ofSize: 17, weight: fontWeight)
+        } else {
+            titleLabel?.font = UIFont.systemFont(ofSize: 17, weight: .bold)
+        }
```
Такая же история, что и с прошлым комментом. 

Еще можно делать [дефолтные значения в объявлении функции](https://www.swiftbysundell.com/articles/default-arguments-in-swift/).

Helvetii951 commented on FakeNFT/Helpers/CustomViews/RateStarImageView.swift:
```diff
@@ -0,0 +1,42 @@
+    func setAppearance(for appearance: Appearance) {
+        switch appearance {
+        case .active:
+//            setImage(UIImage(named: K.Icons.activeStarRate), for: .normal)
+//            isEnabled = false
+            image = UIImage(named: K.Icons.activeStarRate)
+        case .notActive:
+//            setImage(UIImage(named: K.Icons.notActiveStarRate), for: .normal)
+//            isEnabled = false
```
В продакшене лучше убирать неиспользуемый код.

Helvetii951 commented on FakeNFT/Modules/CartModule/Screens/CartMainScreen/ViewModel/CartViewModel.swift:
```diff
@@ -0,0 +1,72 @@
+import Combine
+
+final class CartViewModel {
+    private var cancellables = Set<AnyCancellable>()
+    
+    @Published private (set) var visibleRows: [NftSingleCollection] = []
+    @Published private (set) var cartError: Error?
+    @Published private (set) var requestResult: RequestResult?
```
За использование Combine плюсую👍

Helvetii951 commented on FakeNFT/Modules/CartModule/Screens/CartPaymentResultScreen/Model/RequestResult.swift:
```diff
@@ -0,0 +1,51 @@
+    var description: String {
+        switch self {
+        case .success:
+            return NSLocalizedString("Успех! Оплата прошла,\n поздравляем с покупкой!", comment: "")
+        case .failure:
+            return NSLocalizedString("Упс! Что-то пошло не так :(\n Попробуйте ещё раз!", comment: "")
+        case .loading:
+            return NSLocalizedString("Загружаем данные, пожалуйста, ожидайте", comment: "")
```
Файлов с локализованными строками в проекте нет.

Рекомендую подобные строки называть так:

```
module_name.screen_name.exact_string
```

FierySasquatch168 committed:
added collectionView diffableDataSource to CatalogModule
excluded DefaultNetworkClient initialization from ModuleCoordinator, DefaultNetworkClient now is initialized separately for each screen in ModulesFactory

FierySasquatch168 committed:
added CatalogCollectionViewCell
added image loading for the cell
added layout of the cell
corrected RequestConstructor method to load one nft at a time

FierySasquatch168 committed:
fixed collectionViewCell layout for different devices

FierySasquatch168 committed:
SingleNfts from each collection saved to dataStore
after collection pickUp catalogCollectionViewModel checks if there are stored items in dataStore with equal ids. If there's no data - viewModel commands newtworkClient to load them
SingleNfts stored object changed to Set<SingleNfts> in order to avoid data duplication.

FierySasquatch168 committed:
Merge pull request #3 from FierySasquatch168/storingOfCollectionNfts

SingleNfts from each collection saved to dataStore

FierySasquatch168 committed:
fixed some issues with memory usage (forgotten [weak self])

FierySasquatch168 committed:
catalog and cart interaction added
CartViewModel method 'load()' replaced by 'reload()' and is triggered only if an error occur while loading the cart

FierySasquatch168 committed:
CatalogCollecitonViewController:
new model for visible nfts introduced - with two additional properties "isStored" and "isLiked"
nft now is added to cart by tapping the cart button
tapping the card button updates the cellViewModel
cell UI updates accordingly
when item is deleted from the cart, cellViewModel and cellUI are updated accordingly

FierySasquatch168 committed:
urlCreation logic moved from ViewCells to CellViewModels
like logic implementation introduced

FierySasquatch168 committed:
like functionality added

FierySasquatch168 committed:
universal webView is used for content loading
type of webView data is set up in WebViewUrlSource enum
author link in CatalogCollectionViewController is set after collectionData received
constructSort in AlertConstructor replaced by constructAlert method with parameters
alert actions to alerts are now added by one method addLoadErrorAlertActions

FierySasquatch168 committed:
added loading animation to catalogViewController

FierySasquatch168 committed:
loading indicator in CatalogCollectionViewController added

FierySasquatch168 committed:
fixed sortAlert message

FierySasquatch168 committed:
generic method "func addSortAlertActions<T: CaseIterable & Sortable>(for alert: UIAlertController, values: [T], completion: @escaping (T) -> Void)" in AlertConstructor replaced two methods of separate protocol CartAlertConstructable and CatalogAlertConstructuble
Sortable protocol added to sorting enums to provide T-type generic method

FierySasquatch168 committed:
replaced diffable data sources for tableView and collectionView by generic ones
replaced CartDataSourceManagerProtocol and CatalogDataSourceManagerProtocol by more short and clear GenericTableViewDataSourceProtocol and TableViewDataSourceCoordinatable
replaced PaymentMethodDSManagerProtocol and NftCollectionDSManagerProtocol by more short and clear GenericDataSourceManagerProtocol and CollectionViewDataSourceCoordinatable

FierySasquatch168 committed:
Merge pull request #5 from FierySasquatch168/genericDiffableDataSource

replaced diffable data sources for tableView and collectionView by generic ones

FierySasquatch168 committed:
removed old comments

FierySasquatch168 committed:
remove commonCollectionDataSourceItem due to its unnessessariness

FierySasquatch168 committed:
removed commented code

FierySasquatch168 committed:
Merge pull request #4 from FierySasquatch168/catalog_sprint19

Catalog sprint19
